### PR TITLE
Tests book category dev

### PR DESCRIPTION
**Pull request by don-bigdad**

### CODE REVIEW

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @BeforeAll
+    static void beforeAll(@Autowired WebApplicationContext applicationContext,
+                          @Autowired DataSource dataSource) throws SQLException {
+        mockMvc = MockMvcBuilders.webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+        tearDown(dataSource);
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]: Do we need it here if you tearDown(dataSource) in BeforeEach and AfterEach?

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    @BeforeEach
+    void beforeEach(@Autowired DataSource dataSource) throws SQLException {
+        tearDown(dataSource);
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:55]: Do we need it here if you use tearDown(dataSource) in AfterEach ?

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    void createNewBookAndSaveIntoDbAssertSuccess() throws Exception {
+        CreateBookRequestDto createBookRequestDto = new CreateBookRequestDto(
+                "Book1",
+                "Author1",
+                "ISBN1",
+                BigDecimal.TEN,
+                "Description1",
+                "coverImageUrl1",
+                List.of(1L));
+        BookDto expectedDto = getExpectDto(createBookRequestDto);
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]: You don't need to create expectedDto here.
In the end of the test you should compare createBookRequestDto with actualDto

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+                .content(jsonRequest)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isCreated())
+                .andReturn();
+
+        BookDto actualDto = objectMapper.readValue(result.getResponse().getContentAsString(),
+                BookDto.class);
+        assertNotNull(actualDto);
+        EqualsBuilder.reflectionEquals(expectedDto, actualDto,"id");
+    }
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:124]: Let's add test case when request Dto is invalid and try to create new book

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+    @Test
+    @DisplayName("Get all books")
+    @WithMockUser(username = "user")
+    void getAllBooksFromDbAssertSuccess() throws Exception {
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]:
```suggestion
    void getAll_ListOfBooks_Ok() throws Exception {
```

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+                    new ClassPathResource("database/books/clear-books-db.sql"));
+        }
+    }
+
+    @Test
+    @DisplayName("Create a new book")
+    @Sql(scripts = "classpath:database/books/clear-books-db.sql",
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    void createNewBookAndSaveIntoDbAssertSuccess() throws Exception {
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]:
```suggestion
    void createBook_ValidRequestDto_Ok() throws Exception {
```

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    @WithMockUser(username = "user")
+    void getAllBooksFromDbAssertSuccess() throws Exception {
+        List<BookDto> expectBookDtoList = getBookDtoList();
+
+        MvcResult result = mockMvc.perform(get("/books")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        BookDto[] actualDtoList = objectMapper.readValue(result.getResponse().getContentAsString(),
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]: To get List<BookDto> is better to use 
```suggestion
        List<BookDto> actualDtoList = objectMapper.readValue(result.getResponse().getContentAsString(), 
        new TypeReference<List<BookDto>>() {});
```

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        for (int i = 0; i < actualDtoList.length - 1; i++) {
+            EqualsBuilder.reflectionEquals(expectBookDtoList.get(i),
+                    actualDtoList[i], "categoryIds");
+        }
+    }
+
+    @Test
+    @DisplayName("Get book by id")
+    @WithMockUser(username = "user")
+    void getBookByIdFromDbAssertSuccess() throws Exception {
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]:
```suggestion
    void getBookById_ValidId_Ok() throws Exception {
```

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        BookDto expectBookDto = getBookDtoList().get(1);
+
+        MvcResult result = mockMvc.perform(get("/books/2")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        BookDto actualDtoList = objectMapper.readValue(result.getResponse().getContentAsString(),
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]:
```suggestion
        BookDto actualDto = objectMapper.readValue(result.getResponse().getContentAsString(),
```

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        assertNotNull(actualDtoList);
+        EqualsBuilder.reflectionEquals(expectBookDto,
+                    actualDtoList, "categoryIds");
+    }
+
+    @Test
+    @DisplayName("Delete book by id")
+    @WithMockUser(username = "admin", roles = {"ADMIN","USER"})
+    void deleteBookByIdAssertSuccess() throws Exception {
+
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]:
```suggestion
```

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+        mockMvc.perform(delete("/books/2")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNotFound());
+
+        MvcResult result = mockMvc.perform(get("/books")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        BookDto[] actualDtoList = objectMapper.readValue(result.getResponse().getContentAsString(),
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java]: The same with TypeReference

File: src/test/java/com/example/springbootbookshop/controller/CategoryControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    @Test
+    @DisplayName("Create a new category")
+    @Sql(scripts = "classpath:database/category/clear-category.sql",
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    void createNewCategoryAndSaveIntoDbAssertSuccess() throws Exception {
+        CategoryRequestDto categoryRequestDto = new CategoryRequestDto(
+                "Literature",
+                "Category with different stories");
+        CategoryDto expectedDto = getExpectDto();
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/CategoryControllerTest.java]: the same as book controller test

File: src/test/java/com/example/springbootbookshop/controller/CategoryControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    @WithMockUser(username = "user")
+    void getAllCategoriesFromDbAssertSuccess() throws Exception {
+        List<CategoryDto> expectedCategoryDtoList = getCategoriesList();
+
+        MvcResult result = mockMvc.perform(get("/category")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        CategoryDto[] actualDtoList = objectMapper.readValue(result.getResponse()
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/controller/CategoryControllerTest.java]: the same as book controller test

File: src/test/java/com/example/springbootbookshop/service/BookServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    private CreateBookRequestDto getCreateRequestBookDto() {
+        return new CreateBookRequestDto(
+                "New Title",
+                "New Author",
+                "New ISBN",
+                BigDecimal.valueOf(200),
+                "New Description",
+                "New image url",
+                List.of(1L));
+
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/springbootbookshop/service/BookServiceTest.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/example/springbootbookshop/controller/BookController.java [hunk 15]
```diff
@@ -44,7 +44,7 @@ public List<BookDto> getAll(@PageableDefault(size = 5, page = 0)
     @DeleteMapping("/{id}")
     @PreAuthorize("hasRole('ADMIN')")
     @Operation(summary = "Delete book by id")
-    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/controller/BookController.java:47]: why Not found? it different status code groups))

File: src/main/java/com/example/springbootbookshop/security/CustomUserDetailsService.java [hunk 16]
```diff
@@ -15,6 +15,6 @@ public class CustomUserDetailsService implements UserDetailsService {
     @Override
     public UserDetails loadUserByUsername(String username) {
         return userRepository.findByEmail(username).orElseThrow(() ->
-                new UsernameNotFoundException("User doesn`t exist"));
+                new UsernameNotFoundException("User does`t exist"));
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/security/CustomUserDetailsService.java:18]: Let’s  correct mistake  ;)

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java (hunk 2 above)
Comment (liliia-ponomarenko) [reply src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:55]: not fixed

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+                    new ClassPathResource("database/category/clear-category.sql"));
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource("database/books/clear-books-db.sql"));
+        }
+    }
+
+    @Test
+    @DisplayName("Create a new book")
+    @Sql(scripts = "classpath:database/books/clear-books-db.sql",
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:79]: you should execute it after the test

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+                BookDto.class);
+        assertNotNull(actualDto);
+        EqualsBuilder.reflectionEquals(expectBookDto,
+                actualDto, "categoryIds");
+    }
+
+    @Test
+    @DisplayName("Delete book by id")
+    @WithMockUser(username = "admin", roles = {"ADMIN","USER"})
+    void deleteBookByIdAssertSuccess() throws Exception {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:167]: Let’s  follow test name standards

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+        mockMvc.perform(delete("/books/2")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNotFound());
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:170]: Let’s  fix status code

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+
+        MvcResult result = mockMvc.perform(get("/books")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:175]: this request doesn’t  have any relation to this test case  (це порушує умову ізольваності тестів, результат цього тесту залежить від того чи правильно працює логіка getAll)

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+
+        List<BookDto> actualDtoList = objectMapper.readValue(result.getResponse()
+                        .getContentAsString(), new TypeReference<>() {});
+        assertEquals(2, actualDtoList.size());
+    }
+
+    @Test
+    @DisplayName("Update book by id")
+    @WithMockUser(username = "admin", roles = {"ADMIN","USER"})
+    void updateBookByIdAssertSuccess() throws Exception {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:185]: the same issue

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+        MvcResult result = mockMvc.perform(get("/books/2")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        BookDto dto = objectMapper.readValue(result.getResponse()
+                        .getContentAsString(), BookDto.class);
+        assertEquals("Book 2", dto.getTitle());
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:192]: the same issue

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+        BookDto actualDto = objectMapper.readValue(updatedData.getResponse().getContentAsString(),
+                BookDto.class);
+        assertEquals("Updated Title", actualDto.getTitle());
+        assertEquals("Updated Author", actualDto.getAuthor());
+    }
+
+    @Test
+    @DisplayName("Get book by non-existing id")
+    @WithMockUser(username = "user")
+    void getBookByNoExitedId() throws Exception {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:216]: the same

File: src/test/java/com/example/springbootbookshop/controller/CategoryControllerTest.java [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+                          @Autowired DataSource dataSource) throws SQLException {
+        mockMvc = MockMvcBuilders.webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+        tearDown(dataSource);
+    }
+
+    @BeforeEach
+    void beforeEach(@Autowired DataSource dataSource) throws SQLException {
+        tearDown(dataSource);
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/CategoryControllerTest.java:58]: why do you call this method here?

File: src/test/java/com/example/springbootbookshop/controller/CategoryControllerTest.java [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.jdbc.Sql;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class CategoryControllerTest {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/controller/CategoryControllerTest.java:42]: Let’s  check all test names and the same issues that in previous class

File: src/test/java/com/example/springbootbookshop/service/BookServiceTest.java [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+        assertEquals(expect.getPrice(),actual.getPrice());
+    }
+
+    @Test
+    @DisplayName("Save book")
+    void saveBook() {
+        CreateBookRequestDto dto = getCreateRequestBookDto();
+        bookService.save(dto);
+        bookService.save(dto);
+        verify(bookRepository, times(2)).save(any());
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/springbootbookshop/service/BookServiceTest.java:64]: you can check object  that return

File: src/test/resources/database/books-categories/clear-book-category.sql [hunk 27]
```diff
@@ -0,0 +1 @@
+DELETE FROM books_categories;
```
Comment (liliia-ponomarenko) [inline src/test/resources/database/books-categories/clear-book-category.sql:1]: add one empty line (check other files)

File: src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java (hunk 20 above)
Comment (don-bigdad) [reply src/test/java/com/example/springbootbookshop/controller/BookControllerTest.java:175]: In db i have 3 books and after delete i want to check is the return size of list of db data is 2, its bad test case?

**Pull request merged**